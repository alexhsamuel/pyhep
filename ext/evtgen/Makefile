include ../../standard.mk
include ../../config.mk

ifeq ($(EVTGEN),no)
  $(error EvtGen not configured)
endif

ifeq ($(CERN),no)
  $(error CERNLIB not configured)
endif

.SECONDARY:

#-----------------------------------------------------------------------
# variables
#-----------------------------------------------------------------------

EVTGENINCDIR	= $(EVTGEN)
EVTGENLIBDIR	= $(EVTGEN)

CERNLIBDIR	= $(CERN)/lib
CERNLDFLAGS	= -L$(CERNLIBDIR)

CPPFLAGS	+= -I$(EVTGENINCDIR)
LDFLAGS		+= -L$(CERNLIBDIR)
LDLIBS		+= $(EVTGENLIBDIR)/libEvtGen.a 
LDLIBS		+= $(EVTGENLIBDIR)/libEvtGenModels.a 
LDLIBS		+= $(EVTGENLIBDIR)/libEvtGenBase.a
LDLIBS		+= $(EVTGENLIBDIR)/libCLHEP.a
LDLIBS		+= $(CERNLIBDIR)/libpythia6205.a $(CERNLIBDIR)/libpythiad.a
LDLIBS		+= -ljetset74 -lphotos202 -lpacklib -lmathlib -lg2c 

OBJFILES	= ext.o PyGenerator.o PyParticle.o

# Resolve the parent 'ext.so'.
INSTALLRPATH	= -Wl,-rpath=\$$ORIGIN/.. 

#-----------------------------------------------------------------------
# rules
#-----------------------------------------------------------------------

clean:
	rm -f $(OBJFILES) ext.so 


install:		ext-install.so
	$(INSTALLPROGRAM) -D ext-install.so \
		$(INSTALLLIBDIR)/hep/evtgen/ext.so


setup:
	@ if [ ! -L ../../py/hep/evtgen/ext.so ]; then \
	  ln -s ../../../ext/evtgen/ext.so ../../py/hep/evtgen/ext.so; \
	fi


#-----------------------------------------------------------------------
# dependencies
#-----------------------------------------------------------------------

all:			setup ext.so

PyGenerator.o:		../python.hh PyParticle.hh PyGenerator.hh

PyParticle.o:		../python.hh PyParticle.hh

ext.o:			../python.hh PyParticle.hh PyGenerator.hh

ext.so: LDLIBS += $(PYHEPDIR)/py/hep/ext.so 
ext.so:			$(OBJFILES)

ext-install.so:	LDFLAGS += $(INSTALLRPATH)
ext-install.so: LDLIBS += ext.so
ext-install.so:		$(OBJFILES)

