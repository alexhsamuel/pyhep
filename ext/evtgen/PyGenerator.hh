//----------------------------------------------------------------------
//
// PyGenerator.hh
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

/* Wrapper class for EvtGen generator.  */

#ifndef __PYGENERATOR_HH__
#define __PYGENERATOR_HH__

//----------------------------------------------------------------------
// includes
//----------------------------------------------------------------------

#include <memory>
#include <Python.h>

#include "python.hh"

//----------------------------------------------------------------------
// forward declarations
//----------------------------------------------------------------------

class EvtGen;
class EvtRandomEngine;

//----------------------------------------------------------------------
// classes
//----------------------------------------------------------------------

struct PyGenerator
  : public Py::Object
{
  static PyTypeObject type;
  static PyGenerator* New(const char* decay_file_path, const char* pdt_path);

  PyGenerator(const char* decay_file_path, const char* pdt_path);
  ~PyGenerator();

  std::auto_ptr<EvtRandomEngine> random_engine_;
  std::auto_ptr<EvtGen> generator_;
};


inline PyGenerator*
PyGenerator::New(const char* decay_file_path,
		 const char* pdt_path)
{
  // Construct the Python object.
  PyGenerator* result = Py::allocate<PyGenerator>();
  // Perform C++ construction.
  try {
    new(result) PyGenerator(decay_file_path, pdt_path);
  }
  catch (Py::Exception) {
    Py::deallocate(result);
    throw;
  }
  
  return result;
}


//----------------------------------------------------------------------

#endif  // #ifndef __PYGENERATOR_HH__
