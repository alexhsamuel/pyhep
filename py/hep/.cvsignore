*.pyc
*.pyo
config.py
*-install
pdt.pickle
