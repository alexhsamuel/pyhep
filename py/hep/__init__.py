#-----------------------------------------------------------------------
#
# module hep
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Module for general-purpose code."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from __future__ import generators

import os
import signal
import traceback

#-----------------------------------------------------------------------
# exceptions
#-----------------------------------------------------------------------

class TimeoutException(Exception):
    """A timeout occurred."""

    pass



class NotSupportedError(RuntimeError):
    """A feature is not supported."""

    pass



class SyntaxError(Exception):

    pass



#-----------------------------------------------------------------------
# classes
#-----------------------------------------------------------------------

class Token:

    pass



class Timeout:

    def __timeout_handler(self, signal_number, frame):
        assert signal_number == signal.SIGALRM
        signal.signal(signal.SIGALRM, self.__old_handler)
        self.__active = 0
        raise TimeoutException, self.__message


    def __init__(self, time, message):
        self.__message = message
        self.__old_handler = \
            signal.signal(signal.SIGALRM, self.__timeout_handler)
        self.__active = 1
        signal.alarm(time)


    def cancel(self):
        if not self.__active:
            return
        signal.alarm(0)
        signal.signal(signal.SIGALRM, signal.SIG_DFL)
        self.__active = 0



class MappingAsObject:
    """Implement the object attribute protocol using a mapping.

    A 'MappingAsObject' wraps an object satifying the mapping protocol.
    Attribute references are converted to item gets and sets in the
    underlying mapping."""

    def __init__(self, mapping):
        """Wrap 'mapping' in the object procotol."""
        
        self.__dict__["_MappingAsObject__mapping"] = mapping


    _mapping = property(lambda self: self.__mapping)
    """Read-only access to the underlying mapping."""


    def __getattr__(self, name):
        try:
            return self.__mapping[name]
        except KeyError:
            raise AttributeError, name


    def __setattr__(self, name, value):
        try:
            self.__mapping[name] = value
        except KeyError, exception:
            raise AttributeError, str(exception)



class NrBreitWigner:
    """A non-resonant Breit-Wigner line shape amplitude factor."""

    def __init__(self, mass, width):
        """Construct a non-relativistic Breit-Wigner function.

        'mass' -- The resonance mass.
        
        'width' -- The resonance width."""

        self.__mass = mass
        self.__width = width


    def __call__(self, s):
        """Return the amplitude at energy-squared 's'."""

        mass = self.__mass
        return 1 / (s - complex(mass * mass, mass * self.__width))



#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

try:
    enumerate = __builtins__.enumerate
except AttributeError:
    def enumerate(iterable):
        count = 0
        for item in iter(iterable):
            yield count, item
            count += 1


try:
    sorted = __builtins__.sorted
except AttributeError:
    def sorted(iterable, cmp=cmp):
        result = list(iterable)
        result.sort(cmp)
        return result


try:
    reversed = __builtins__.reversed
except AttributeError:
    def reversed(iterable):
        result = list(iterable)
        result.reverse()
        return result
    

def sort(sequence, sort_fn=None):
    """Return a copy of 'sequence', sorted by 'sort_fn'.

    'sort_fn' -- The comparison function to use.  If 'None', the default
    (built-in 'cmp') is used."""

    sequence = list(sequence)
    if sort_fn is None:
        sort_fn = cmp
    sequence.sort(sort_fn)
    return sequence


def binarySearch(sequence, value, compare=cmp):
    """Find 'value' in sorted 'sequence' by binary search.

    'sequence' -- A sequence of values.  It is assumed to be sorted by
    'compare'.

    'compare' -- A comparison function.

    returns -- The index of the largest element in 'sequence' smaller
    than 'value', or -1 if there is no such element."""

    if len(sequence) == 0:
        return -1
    if compare(value, sequence[0]) < 0:
        return -1
    if compare(value, sequence[-1]) > 0:
        return len(squence) - 1

    i0 = 0
    i2 = len(sequence) - 1
    while True:
        if i2 <= i0 + 1:
            return i2 - 1
        i1 = (i0 + i2) // 2
        if compare(value, sequence[i1]) < 0:
            i2 = i1
        else:
            i0 = i1


def mapFromKeys(key_sequence, value_function):
    """Construct a map from a sequence of keys.

    A map is created whose keys are the elements of 'key_sequence'.  For
    each key, the corresponding value is constructed from the key using
    'value_function'.  If 'sequence' contains a particular key more than
    once, only one is placed in the map (which one is undefined).

    'key_sequence' -- A sequence of key values.

    'value_function' -- A callable for determining the value for each
    element of the sequence.  It is invoked for each key, passing the
    kay as the single argument."""
    
    result = {}
    for key in key_sequence:
        result[key] = value_function(key)
    return result


def mapFromValues(key_function, value_sequence):
    """Construct a map from a sequence of values.

    A map is created whose values are the elements of 'value_sequence'.
    For each value, the corresponding key is constructed from the value
    using 'key_function'.  If 'key_function' assigns the same key to
    more than one element of the sequence, only one element is actually
    used in the map (which one is undefined).

    'key_function' -- A callable for determining the key for each
    element of the sequence.  It is invoked for each value, passing
    the value as the single argument.

    'value_sequence' -- A sequence of values."""
    
    result = {}
    for value in value_sequence:
        result[key_function(value)] = value
    return result


def invertMap(map):
    """Produce a dictionary by inverting keys and values of 'map'."""

    return dict([ (value, key) for key, value in map.items() ])


def remove(dict, key):
    """Get and remove a dictionary item.

    'dict' -- The dictionary from which to get and remove the item.

    'key' -- The key of the item to get and remove.

    returns -- The value corresponding to 'key' in 'dict'.  

    raises -- 'KeyError' if 'dict' has no key 'key'."""
    
    value = dict[key]
    del dict[key]
    return value


def popkey(map, key, default=None):
    """Get and remove a dictionary item.

    'dict' -- The dictionary from which to get and remove the item.

    'key' -- The key of the item to get and remove.

    'default' -- The value to return if 'key' is not in 'map'.

    returns -- The value corresponding to 'key' in 'dict', or 'default'
    if none."""

    if key in map:
        result = map[key]
        del map[key]
        return result
    else:
        return default


def getHostName():
    """Return the name of this host."""

    import socket

    try:
        return socket.gethostbyname_ex(socket.gethostname())[0]
    except socket.error:
        return "(unknown)"


def isMap(object):
    """Determine whether an object satisfies the mapping protocol.

    returns -- True if 'object' satisfies the map protocol as
    specified.""" 

    # FIXME.  There probably is a more consistent way to do this.

    # If the object is an instance of the standard 'dict' class (or a
    # subclass of it), it must satisfy the mapping protocol.
    if isinstance(object, dict):
        return True
    # Check for the required methods.
    for method_name in ("get", "__getitem__", "__len__", ):
        if not hasattr(object, method_name):
            return False
    # The object checks out.
    return 1


def mkdirRecursively(path):
    """Create a directory, including parent directories if necessary."""

    parent, name = os.path.split(path)
    if parent != "" and not os.path.isdir(parent):
        mkdirRecursively(parent)
    os.mkdir(path)


def rmRecursively(path):
    """Remove the file or directory at 'path'.

    If 'path' is a directory, its contents are removed first."""

    if os.path.isdir(path):
        # Construct a list of contents of the directory.
        contents = map(lambda e, path=path: os.path.join(path, e),
                       os.listdir(path))
        # Remove them all.
        map(rmRecursively, contents)
        # Remove the directory itself.
        os.rmdir(path)
    else:
        os.unlink(path)


