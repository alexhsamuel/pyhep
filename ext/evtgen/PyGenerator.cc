//----------------------------------------------------------------------
//
// PyGenerator.hh
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

//----------------------------------------------------------------------
// includes
//----------------------------------------------------------------------

#include <cassert>
#include <string>
#include <unistd.h>

#include "EvtGen/EvtGen.hh"
#include "EvtGenBase/EvtRandom.hh"
#include "EvtGenBase/EvtRandomEngine.hh"
#include "PyGenerator.hh"
#include "PyParticle.hh"
#include "python.hh"
#include "random.hh"

using namespace Py;

//----------------------------------------------------------------------
// classes
//----------------------------------------------------------------------

class EvtShuffledLEcuyerRandomEngine
  : public EvtRandomEngine
{
public:

  virtual ~EvtShuffledLEcuyerRandomEngine();
  virtual double random();

private:

  ShuffledLEcuyerRandom engine_;

};


//----------------------------------------------------------------------
// class members
//----------------------------------------------------------------------

EvtShuffledLEcuyerRandomEngine::~EvtShuffledLEcuyerRandomEngine()
{
}


double 
EvtShuffledLEcuyerRandomEngine::random()
{
  return engine_.random();
}


//----------------------------------------------------------------------

PyGenerator::PyGenerator(const char* decay_file_path,
			 const char* pdt_path)
  : random_engine_(new EvtShuffledLEcuyerRandomEngine)
{
  generator_.reset
    (new EvtGen(decay_file_path, pdt_path, random_engine_.get()));
}


PyGenerator::~PyGenerator()
{
}


//----------------------------------------------------------------------
// Python support
//----------------------------------------------------------------------

namespace {

/* Extract a path from an argument tuple.

   Extracts the specified argument, converts it to a path string, and
   checks that the path is an accessible (for reading) file.

   'args' -- An argument tuple.

   'index' -- The index of the argument to get.

   returns -- The path string.
*/
inline std::string
getArgPath(Tuple* args,
	   int index)
{
  // Extract the path from the argument tuple.
  Ref<Object> arg_obj = args->GetItem(index);
  std::string path = arg_obj->StrAsString();
  // Make sure the file exists and is readable.
  if (access(path.c_str(), R_OK) != 0)
    throw Exception(PyExc_IOError, "cannot open '%s': %s", 
		    path.c_str(), strerror(errno));
  // All good.
  return path;
}


PyObject*
tp_new(PyTypeObject* type,
       Tuple* args,
       Dict* kw_args)
try {
  std::string pdt_path;
  std::string decay_file_path;

  // Determine the path to the particle data file.
  if (args->Size() < 1) {
    // Use the default particle data file.
    Ref<Object> default_pdt_path = 
      import("hep.evtgen", "default_pdt_path");
    pdt_path = default_pdt_path->StrAsString();
  }
  else 
    // The first argument is the path to the particle data file.
    pdt_path = getArgPath(args, 0);

  // Determine the path to the decay file.
  if (args->Size() < 2) {
    // Use the default decay file.
    Ref<Object> default_decay_file_path =
      import("hep.evtgen", "default_decay_file_path");
    decay_file_path = default_decay_file_path->StrAsString();
  }
  else 
    // The second argument is the path to the main decay file.
    decay_file_path = getArgPath(args, 1);

  // Construct the generator.
  Ref<PyGenerator> result = 
    PyGenerator::New(decay_file_path.c_str(), pdt_path.c_str());

  // Interpret any additional arguments as paths to user decay files,
  // and load these.
  for (int a = 2; a < args->Size(); ++a) {
    std::string path = getArgPath(args, a);
    result->generator_->readUDecay(path.c_str());
  }

  // All done.
  return result.release();
}
catch (Exception) {
  return NULL;
}


void
tp_dealloc(PyGenerator* self)
try {
  // Perform C++ deallocation.
  self->~PyGenerator();
  // Free memory for the Python object.
  PyMem_DEL(self);
}
catch (Exception) {
  return;
}


PyObject*
method_decay(PyGenerator* self,
	     Tuple* args)
try {
  PyParticle* particle;
  args->ParseTuple("O!", &PyParticle::type, &particle);

  self->generator_->generateDecay
    (const_cast<EvtParticle*> (particle->particle_));

  RETURN_NONE;
}
catch (Exception) {
  return NULL;
}


PyObject*
method_readDecayFile(PyGenerator* self,
		     Tuple* args)
try {
  const char* decay_file_path;
  args->ParseTuple("s", &decay_file_path);
  // Make sure the file exists and is readable.
  if (access(decay_file_path, R_OK) != 0)
    throw ExceptionFromErrno();
    
  self->generator_->readUDecay(decay_file_path);

  RETURN_NONE;
}
catch (Exception) {
  return NULL;
}


PyMethodDef
tp_methods[] = {
  { "decay", (PyCFunction) method_decay, METH_VARARGS, NULL },
  { NULL, NULL, 0, NULL }
};


}  // anonymous namespace


PyTypeObject
PyGenerator::type = 
{
  PyObject_HEAD_INIT(&PyType_Type)
  0,                                    // ob_size
  "Generator",                          // tp_name
  sizeof(PyGenerator),                  // tp_size
  0,                                    // tp_itemsize
  (destructor) tp_dealloc,              // tp_dealloc
  NULL,                                 // tp_print
  NULL,                                 // tp_getattr
  NULL,                                 // tp_setattr
  NULL,                                 // tp_compare
  NULL,                                 // tp_repr
  NULL,                                 // tp_as_number
  NULL,                                 // tp_as_sequence
  NULL,                                 // tp_as_mapping
  NULL,                                 // tp_hash
  NULL,                                 // tp_call
  NULL,                                 // tp_str
  NULL,                                 // tp_getattro
  NULL,                                 // tp_setattro
  NULL,                                 // tp_as_buffer
  Py_TPFLAGS_DEFAULT
  | Py_TPFLAGS_BASETYPE,                // tp_flags
  NULL,                                 // tp_doc
  NULL,                                 // tp_traverse
  NULL,                                 // tp_clear
  NULL,                                 // tp_richcompare
  0,                                    // tp_weaklistoffset
  NULL,                                 // tp_iter
  NULL,                                 // tp_iternext
  tp_methods,                           // tp_methods
  NULL,                                 // tp_members
  NULL,                                 // tp_getset
  NULL,                                 // tp_base
  NULL,                                 // tp_dict
  NULL,                                 // tp_descr_get
  NULL,                                 // tp_descr_set
  0,                                    // tp_dictoffset
  NULL,                                 // tp_init
  NULL,                                 // tp_alloc
  (newfunc) &tp_new,                    // tp_new
};


//----------------------------------------------------------------------
// global functions
//----------------------------------------------------------------------

extern "C"
{


// These two functions are called by JetSet.

float 
begran_(int*)
{
  return EvtRandom::Flat();
}


float 
rlu_()
{
  return EvtRandom::Flat();
}


}  // extern "C"
