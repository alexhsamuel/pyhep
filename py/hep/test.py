#-----------------------------------------------------------------------
#
# module hep.test
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Regression testing support."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from   hep.bool import *
import sys

#-----------------------------------------------------------------------
# exceptions
#-----------------------------------------------------------------------

class TestFailure(Exception):
    pass



#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

def assert_(expression):
    if not expression:
        raise TestFailure, "assertion failed"


def compare(value, expected, precision=None):
    if type(value) in (list, tuple) and type(expected) in (list, tuple):
        if len(value) != len(expected):
            result = False
        else:
            try:
                result = reduce(
                    lambda r1, r2: r1 and r2,
                    map(lambda (v, e): compare(v, e, precision),
                        zip(value, expected)),
                    True)
            except TestFailure:
                result = False

    elif precision is not None:
        result = (abs(value - expected) <= precision)

    else:
        result = (value == expected)

    if result:
        return True
    else:
        raise TestFailure, \
              "expected %s; got %s" % (repr(expected), repr(value))


def compareSequence(value, expected, ordered=True):
    value = list(value)
    expected = list(expected)
    if not ordered:
        value.sort()
        expected.sort()
    return compare(value, expected)


