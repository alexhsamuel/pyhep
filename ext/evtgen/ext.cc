//----------------------------------------------------------------------
//
// evtgen/ext.cc
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

/* Python extension module for access to EvtGen libraries.  */

//----------------------------------------------------------------------
// includes
//----------------------------------------------------------------------

#include <cassert>
#include <iostream>
#include <Python.h>

#include "PyParticle.hh"
#include "PyGenerator.hh"
#include "python.hh"

using namespace Py;

//----------------------------------------------------------------------
// Python module definition
//----------------------------------------------------------------------

namespace {

PyMethodDef
functions[] = 
{
  { NULL, NULL, 0, NULL }
};

PyTypeObject*
types[] = {
  &PyGenerator::type,
  &PyParticle::type,
  NULL
};


}  // anonymous namespace


extern "C" {

void 
initext()
try {
  Module::Initialize("ext", functions, types, "");
}
catch (Exception) {
  std::cerr << "could not initialize module 'ext'" << std::endl;
}


}  // extern "C"
