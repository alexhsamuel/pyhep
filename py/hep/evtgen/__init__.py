#-----------------------------------------------------------------------
#
# module hep.evtgen
#
# Copyright 2004 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""PyHEP interface to the EvtGen decay generation package."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from   ext import Generator, Particle
import hep.config
import os.path

#-----------------------------------------------------------------------
# constants
#-----------------------------------------------------------------------

default_pdt_path = os.path.join(hep.config.data_dir, "pdt.dat")
default_decay_file_path = os.path.join(hep.config.data_dir, "decay.dat")

