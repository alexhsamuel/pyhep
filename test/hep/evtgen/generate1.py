"""Test of basic decay-generation functionality."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

import hep.evtgen
from   hep.test import compare, assert_

#-----------------------------------------------------------------------
# constants
#-----------------------------------------------------------------------

num_decays = 100

#-----------------------------------------------------------------------
# test
#-----------------------------------------------------------------------

generator = hep.evtgen.Generator()

for i in xrange(num_decays):
    # Generate a rho0 decay.
    parent = hep.evtgen.Particle("rho0")
    generator.decay(parent)
    # Make sure it's still a rho0.
    compare(parent.species, "rho0")
    # Check that it decayed to pi+ pi-.
    decay_species = [ c.species for c in parent.decay_products ]
    compare(decay_species, [ "pi+", "pi-" ])

