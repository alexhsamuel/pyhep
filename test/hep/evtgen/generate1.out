Traceback (most recent call last):
  File "generate1.py", line 7, in ?
    import hep.evtgen
  File "/home/samuel/dev/pyhep/py/hep/evtgen/__init__.py", line 15, in ?
    from   ext import Generator, Particle
ImportError: No module named ext
